content/post/code-as-configuration.md

```diff
@@ -0,0 +1,112 @@
+
+Some readers may note that this is not actually a novel idea: many companies, especially the really large tech companies (FAANG et al.), invest a lot of effort in making sure their software is architected following this paradigm so that their specialists can be specialized. However, in my work with smaller companies I've often encountered an unfamiliarity with or reluctance to utilize this design pattern. In this post I want to make the case that Software Engineering teams should pro-actively look for opportunities to pursue the *code as configuration* paradigm in order to:
+
+* Unblock their Analytics team
+* Improve iteration speed
+* Promote specialization (so analysts and engineers can work on what they are independently good at)
+* Remove tedious business-logic tasks from the Software Engineering backlog
+
+## The Problem
+
+In my work as a data-scientist over the last 10 years, I have written *a lot* of code. I would wager that the majority of my working time over that period has been spent writing and editing code. Unfortunately, most of that code is, from a software engineering perspective, "bad". Like many folks working in the data fields, I'm a [self taught](https://xkcd.com/1513/) programmer, and, frankly, writing readable, DRY, efficient code has rarely been a priority for me ("it only has to run once" is a common if frightening refrain among academics and statisticians).
+
+Caveats aside, I consider myself a *deeply technical* person, but from the vantage of the software engineers I work with, I occupy a bizarre middle-ground: I am technical enough that I can engage with them in architecture discussions, and I know how to use vim, but I very clearly cannot be trusted to write customer-facing production code on my own. Additionally, there is a lot of background knowledge that software engineers are assumed to have like "how the internet works" that I simply do not possess. Occupying this middle-ground of "oddly technical, yet does not understand things" can make it hard to partner and communicate with the broader software engineering organization.
+
+After spending a lot of time collaborating with software engineers with varying degrees of success, I have determined that the optimal pattern for collaboration relies on architecting and  building systems where I (and the other data folks on my team) can write and deploy code / scripts without:
+
+* needing to get that code approved by software engineers
```

Comment:
I'm sort of eliding some complexity here because I don't want to get into the weeds of what a "production environment" is. 

Phrases like "the application production environment" assume that there is *one* production application. I'd make the claim that this framing is misleading, especially in data work -- I consider the Looker environment used to support the business *an* application production environment. I'd consider the production DBT code *an* application production environment. If there's a model that writes predictions into a redis store that is queried by the customer-facing application (but is fully decoupled via the redis API), is that *the* application production environment, or just another application (presumably with multiple environments as well)?

At most companies there are many different applications, each of which has different environments (local, testing, staging, prod). People may want to (and should!) put different processes and control around those different applications. I don't really want to get into that in this post although maybe it deserves a post of its own.

Does that make sense? Do you think I need to clarify this more in the post? (If this thought is going to be distracting to readers than it may be worth doing a "what is production" post before I do this one)

Reply:
I agree that there are multiple application environments especially as the complexity of the organization's offering grows. I suppose what I'm more concerned with in regards to this comment is who owns the environment in question? I can see multiple levels of ownership:

1. The Looker environment is completely owned by Data, no sign off required by SE
2. My ML model that makes API calls to our main customer facing application needs SE approval for certain tasks, but that mostly takes the form of ensuring I don't break or slow things down. 
3. Any real code change I would like to make to the application will certainly be reviewed heavily.

And there might be more levels still. I'm okay with leaving out any of this detail for the sake of clarity, but I can see some Data folks who have never set foot inside the main application production environment get perhaps the wrong idea, namely that they are able to write application code if the SE team creates the right environment for them.